Make sure circuits with constant ops can be serialized to JSON correctly

yuyichao opened the pull request:
The fix is limited to OpConst to make sure we can handle the deserialization without any problems. Also fixed the implementation of ShiftConst to handle BigInt produced by constant folding.

Add tests to check the validity and stability of the JSON object.

Fix #65

.... took me longer to add the tests than to fix the issue....

tilk commented on src/cells/arith.mjs:
```diff
@@ -368,6 +368,13 @@ export const OpConst = Arith.define('OpConst', {
+        let constant = this.get('constant');
+        if (typeof constant === 'string' || constant instanceof String) {
+            constant = BigInt(constant);
+            if (constant <= BigInt(Number.MAX_SAFE_INTEGER))
```
What about negative numbers?

tilk commented on src/cells/arith.mjs:
```diff
@@ -384,6 +391,19 @@ export const OpConst = Arith.define('OpConst', {
+        const params = Arith.prototype.getGateParams.apply(this, arguments);
+        // `constant` may be a bigint and we need to encode it manually
+        if (typeof params.constant === 'bigint') {
+            if (params.constant <= BigInt(Number.MAX_SAFE_INTEGER)) {
```
Negative numbers.

tilk commented on tests/index.test.mjs:
```diff
@@ -366,6 +476,34 @@ describe.each([
+                }
+            };
+            // Make sure we test both Number and BigInt
+            testcon(randInt(0, Math.min(2**bits - 1, Number.MAX_SAFE_INTEGER)));
```
Negative numbers are not tested.

tilk reviewed (changes requested):
One more comment: if you look at `integrateArithConstant`, you'll see it only generates *Const cells for "small" constants - which was done for diagram readability reasons. But the constant is always of BigInt type, and it could already be a Number there.

yuyichao commented:
Rebased and fixed.

This now eagerly convert the constant to `Number` when doing the transformation but it will still accept bigint if explicitly constructed.
